traitlets/traitlets.py

```diff
@@ -421,6 +429,27 @@ def instance_init(self, obj):
         """
         pass
 
+    def _lineage(self):
+        parent = self
+        yield parent
+        while parent._parent is not None:
+            parent = parent._parent()
+            yield parent
+
+    def __str__(self):
+        if self.this_class is not None:
+            lineage = list(self._lineage())
+            absolute_name = lineage[-1].name
```

Comment:
If this is the only thing the parent tree is used for and the name is always `lineage[-1].name`, why isn't `name` just passed down the lineage instead of recording the lineage itself?

Reply:
@minrk we use the lineage in the next line down to indicate how traits are nested.

If we don't do this `obj.attr = (1, 2, 3)` for a `Tuple(String())` will give an error like:

```
The 'attr' trait of MyClass instances expected a string, but got an int instead.
```

This is quit confusing since the bad value is not the tuple, but the int inside the tuple.

Reply:
If parent is truly needed, the other option is to do attribute assignment, rather than an API change:

```python
trait.parent = self
trait.class_init(obj, name)
```

Reply:
Now that you are on it, to me the message:

    The 'attr' trait of MyClass instances expected ..,

is disorientating.  It would be faster to grasp if it said:

    The `MyClass.trait` trait expected ...